engine/Program.js

```diff
@@ -1,37 +1,35 @@
-(function() {
-  "use strict";
-
-  window.Program = function(gl, vertexShader, fragmentShader) {
+class Program {
+  constructor(gl, vertexShader, fragmentShader) {
+    this.gl = gl;
     this.vertexShader = vertexShader;
     this.fragmentShader = fragmentShader;
-    this.program = gl.createProgram();
-
-    this.initialize = function() {
-      gl.attachShader(this.program, this.vertexShader.shader);
-      gl.attachShader(this.program, this.fragmentShader.shader);
-      gl.linkProgram(this.program);
+    this.program = this.gl.createProgram();
+    this.setupProgram();
+  }
 
-      if (!gl.getProgramParameter(this.program, gl.LINK_STATUS)) {
-        var programLog = gl.getProgramInfoLog(this.program);
-        this.tearDown();
-        throw new Error(programLog);
-      }
-    };
+  bind() {
+    this.gl.useProgram(this.program);
+  };
 
-    this.bind = function() {
-      gl.useProgram(this.program);
-    };
+  unbind() {
+    this.gl.useProgram(null);
+  };
 
-    this.unbind = function() {
-      gl.useProgram(null);
-    };
+  tearDown() {
+    this.gl.deleteProgram(this.program);
+    this.gl.deleteShader(this.vertexShader);
+    this.gl.deleteShader(this.fragmentShader);
+  };
 
-    this.tearDown = function() {
-      gl.deleteProgram(this.program);
-      gl.deleteShader(this.vertexShader.shader);
-      gl.deleteShader(this.fragmentShader.shader);
-    };
+  setupProgram() {
```

Comment:
Why did you extract this into a setup function? Wouldn't it fit directly in the constructor, like in the `Mesh` or `Buffer` class?

Reply:
I had it like that first. But since we have a `tearDown`, I thought I stick to the pattern.